File: erlang/customer.erl
-module(customer).
-export([run/2]).

run(Name, Orders) ->
  Orders ! {place, {self(), "soup"}},
  eat(Name, Orders, 10).

eat(Name, _, 0) ->
  io:format("~s going home~n", [Name]);
eat(Name, Orders, Times) ->
  receive
    {serve, Meal} ->
      io:format("~s eating ~p~n", [Name, Meal]),
      timer:sleep(rand:uniform(5000)),
      Orders ! {place, {self(), "soup"}},
      eat(Name, Orders, Times - 1)
  after 7000 ->
    io:format("~s waiting too long, going shopping~n", [Name]),
    timer:sleep(rand:uniform(5000)),
    eat(Name, Orders, Times)
  end.

File: go/restaurant/customer.go
package main

import (
	"sync"
	"time"
)

func Customer(name string, wg *sync.WaitGroup) {
	defer wg.Done()
	ch := make(chan *Order)
	for mealsEaten := 0; mealsEaten < 10; {
		select {
		case Orders <- &Order{food: "soup", customer: name, reply: ch}:
			meal := <-ch
			do(name, "eating " + meal.food + " prepeared by " +
				meal.preparedBy, 10E9)
			mealsEaten += 1
		case <-time.After(time.Second * 7):
			do(name, "waiting too long, going shopping", 5E9)
		}
	}
	report(name, "going home")
}
